### PR TITLE
Type Providers: invalidate daemon for projects with generative provided types

### PR DESCRIPTION
**Pull request by DedSec256**
This PR enables the daemon invalidation for cases when, after building a project with generative type providers, it is necessary to reanalyze projects that refer to it on the fly.

This PR also contains a fix in the propagation logic of ProvidedTypeContext between the generative provided types. At the moment, we are caching the provided types, and any ProvidedTypeContext's that FCS can apply to them, which can lead to errors. The fix suggests creating lightweight wrappers containing ProvidedTypeContext for cached provided entities.

JetBrains1

### CODE REVIEW

**Comment by DedSec256** [comment]
We should also check this branch on our internal CI

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 1]
```diff
@@ -17,6 +18,7 @@ type IProxyExtensionTypingProvider =
     inherit IExtensionTypingProvider
 
     abstract RuntimeVersion: unit -> string
+    abstract HasGenerativeTypeProviders: assemblies: string seq -> bool
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:21]:
```suggestion
    abstract HasGenerativeTypeProviders: outputPaths: string seq -> bool

```

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 2]
```diff
@@ -114,5 +116,11 @@ type ExtensionTypingProviderShim(solution: ISolution, toolset: ISolutionToolset,
 
             $"{inProcessDump}\n\n{outOfProcessDump}"
 
+        member this.HasGenerativeTypeProviders(assemblies) =
+            // We can determine which projects contain generative provided types
+            // only from type providers hosted out-of-process
+            not (isConnectionAlive()) ||
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:122]: Why should it be `true` when the connection is not alive?

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs [hunk 3]
```diff
@@ -315,15 +320,26 @@ type FcsProjectProvider(lifetime: Lifetime, solution: ISolution, changeManager:
...
+        typeProvidersShim: IProxyExtensionTypingProvider) =
     do
         scheduler.EnqueueTask(SolutionLoadTask("FSharpProjectOptionsProvider", SolutionLoadTaskKinds.StartPsi, fun _ ->
             // todo: track file system changes instead? This currently may be triggered on a project model change too.
             outputAssemblies.ProjectOutputAssembliesChanged.Advise(lifetime, fun (project: IProject) ->
-                if not fcsProjectProvider.HasFcsProjects || project.IsFSharp then () else
+                if not fcsProjectProvider.HasFcsProjects ||
+                   project.IsFSharp &&
+                   fcsProjectProvider.GetProjectOutputPaths(project)
+                   |> typeProvidersShim.HasGenerativeTypeProviders |> not then () else
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs:342]: Would it be hard to cache it? We don't want it to be repeatedly called on big solutions.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs [hunk 4]
```diff
@@ -18,40 +18,45 @@ type internal TypeProvidersCache() =
...
+        addTypeProvider projectAssembly tpAssembly tp
 
     member x.Get(id) =
         let hasValue = SpinWait.SpinUntil((fun () -> proxyTypeProvidersPerId.ContainsKey id), 15_000)
 
         if not hasValue then failwith $"Cannot get type provider {id} from TypeProvidersCache"
         else proxyTypeProvidersPerId.[id]
 
+    member x.Get(assembly) =
+        match typeProvidersPerAssembly.TryGetValue assembly with
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:56]: Why some single-arg applications in this PR have parens and some do not?

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs [hunk 5]
```diff
@@ -74,13 +79,14 @@ type IProxyTypeProvidersManager =
...
 
+    abstract member HasGenerativeTypeProviders: assembly: string -> bool
     abstract member Dump: unit -> string
 
 type TypeProvidersManager(connection: TypeProvidersConnection) =
     let protocol = connection.ProtocolModel.RdTypeProviderProcessModel
     let lifetime = connection.Lifetime
-    let tpContext = TypeProvidersContext(connection)
     let typeProviders = TypeProvidersCache()
+    let tpContext = TypeProvidersContext(connection)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:89]: Just for reference: does it change something?

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs (hunk 2 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:122]: in case out-of-process type-providers are disabled, then the general check should be equivalent to a simple `project.IsFSharp` check. You seem to be right, there should be isConnectionAlive () && ...

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs [hunk 6]
```diff
@@ -74,23 +82,31 @@ type IProxyTypeProvidersManager =
...
+
+    do
+        connection.Execute(fun () ->
+            protocol.Invalidate.Advise(lifetime, fun id -> typeProviders.Get(id).OnInvalidate()))
 
-    do connection.Execute(fun () ->
-        protocol.Invalidate.Advise(lifetime, fun id -> typeProviders.Get(id).OnInvalidate()))
+        fcsProjectProvider.ModuleInvalidated.Advise(lifetime, fun psiModule ->
+            let project = getModuleProject psiModule
+            if isNull project then () else
```
Comment (DedSec256) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:101]: some special case, it would be better to do assert here

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs (hunk 3 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs]: implemented memoization of calculations

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs [hunk 7]
```diff
@@ -74,23 +83,30 @@ type IProxyTypeProvidersManager =
...
+    abstract member HasGenerativeTypeProviders: project: IProject -> bool
     abstract member Dump: unit -> string
 
-type TypeProvidersManager(connection: TypeProvidersConnection) =
+type TypeProvidersManager(connection: TypeProvidersConnection, fcsProjectProvider: IFcsProjectProvider) =
     let protocol = connection.ProtocolModel.RdTypeProviderProcessModel
     let lifetime = connection.Lifetime
     let tpContext = TypeProvidersContext(connection)
     let typeProviders = TypeProvidersCache()
+    let projectHasGenerativeTypeProvider = Dictionary<IProject, bool>()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:94]: `projectsWithGenerativeProviders`, `projectsContainingGenerativeProviders`, `projectsReferencingGenerativeProviders`?

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsCheckerService.fs [hunk 8]
```diff
@@ -25,6 +25,34 @@ open JetBrains.ReSharper.Psi.Tree
 open JetBrains.ReSharper.Psi.VB
 open JetBrains.Util
 
+[<AutoOpen>]
+module FcsProjectProviderUtils =
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsCheckerService.fs:29]: Why this module was moved to `FcsCheckerService.fs`?

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsCheckerService.fs [hunk 9]
```diff
@@ -25,6 +25,34 @@ open JetBrains.ReSharper.Psi.Tree
...
+        | :? IProject as project -> project.IsFSharp // todo: check `isOpened`?
+        | _ -> false
+
+    let getModuleProject (psiModule: IPsiModule) =
+        psiModule.ContainingProjectModule.As<IProject>()
+
+    let isFSharpProjectModule (psiModule: IPsiModule) =
+        psiModule.IsValid() && isFSharpProject psiModule.ContainingProjectModule // todo: remove isValid check?
+
+    let [<Literal>] invalidateProjectChangeType =
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsCheckerService.fs:47]: How is this relevant to the FCS checker?

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs [hunk 10]
```diff
@@ -315,15 +287,24 @@ type FcsProjectProvider(lifetime: Lifetime, solution: ISolution, changeManager:
         member this.GetFcsProject(psiModule) =
             getOrCreateFcsProject psiModule
 
+        member this.GetProjectOutputPaths(project) =
+            seq { for psiModule in projectsPsiModules.[project] do
+                  match tryGetValue psiModule fcsProjects with
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsProjectProvider.fs:292]: Please fix the indentation.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs [hunk 11]
```diff
@@ -13,45 +17,51 @@ open JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol.Models
...
     member x.Get(id) =
         let hasValue = SpinWait.SpinUntil((fun () -> proxyTypeProvidersPerId.ContainsKey id), 15_000)
 
         if not hasValue then failwith $"Cannot get type provider {id} from TypeProvidersCache"
         else proxyTypeProvidersPerId.[id]
 
+    member x.Get(outputPath) =
+        match typeProvidersPerAssembly.TryGetValue(outputPath) with
+        | true, x -> x.Values
+        | false, _ -> [||] :> _
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:63]: `[||]` is not currently a singleton and will needlessly allocate. Let's use the empty array instance from `JetBrains.Util` or at least BCL?

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs [hunk 12]
```diff
@@ -74,23 +84,33 @@ type IProxyTypeProvidersManager =
...
+        connection.Execute(fun () ->
+            protocol.Invalidate.Advise(lifetime, fun id -> typeProviders.Get(id).OnInvalidate()))
 
-    do connection.Execute(fun () ->
-        protocol.Invalidate.Advise(lifetime, fun id -> typeProviders.Get(id).OnInvalidate()))
+        fcsProjectProvider.ModuleInvalidated.Advise(lifetime, fun psiModule ->
+            let project = getModuleProject psiModule
+            try lock.Enter()
+                doesProjectContainGenerativeProviders.Remove(project.NotNull()) |> ignore
+            finally lock.Exit())
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:106]: Could this and other `lock` usages be simpler with `JetFastSemiReenterableRWLock`?

File: ReSharper.FSharp/test/src/FSharp.Common.Tests/SciptPsiModulesTest.fs [hunk 13]
```diff
@@ -128,3 +129,15 @@ type FSharpFileServiceStub() =
     interface IFSharpFileService with
         member x.IsScratchFile _ = false
         member x.IsScriptLike _ = false
+
+[<SolutionComponent>]
+type TestExtensionTypingProviderStub() =
```
Comment (auduchinok) [inline ReSharper.FSharp/test/src/FSharp.Common.Tests/SciptPsiModulesTest.fs:134]: Would it be hard to have a single `TestExtensionTypingProviderStub` component? Do I get it right that there's a zoning issue? Ideally we should be trying to remove these stubs instead of adding more. 🙂

Comment (DedSec256) [reply ReSharper.FSharp/test/src/FSharp.Common.Tests/SciptPsiModulesTest.fs:134]: > Do I get it right that there's a zoning issue?
Exactly

File: ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsCheckerService.fs (hunk 8 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsCheckerService.fs:29]: I needed the `getModuleProject` function, so I moved the module containing it closer to IFcsProjectProvider

Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Checker/FcsCheckerService.fs:29]: What do you think is the best thing to do?

File: ReSharper.FSharp/test/src/FSharp.Common.Tests/SciptPsiModulesTest.fs (hunk 13 above)
Comment (auduchinok) [reply ReSharper.FSharp/test/src/FSharp.Common.Tests/SciptPsiModulesTest.fs:134]: What components are missing?

**Review by auduchinok — APPROVED** [review]
